session_4/blog/app/controllers/articles_controller.rb

```diff
@@ -0,0 +1,108 @@
+class ArticlesController < ApplicationController
+  load_and_authorize_resource
+
+  before_action :set_article, only: [:show, :edit, :update, :destroy]
+
+  # GET /articles
+  # GET /articles.json
+  def index
+    #to show popular articles
+    @popular_articles=Article.order(views: :DESC).limit(3)
+    #unregistered users can only read 5 articles
+    if session[:user_id] && !session[:private_articles_left]
+      session[:private_articles_left]=current_user.private_articles_left
+    end
+    if !session[:user_id]
+      flash[:notice] ||="Login to read private article"
+    end
+    @articles = Article.all
+  end
+
+  # GET /articles/1
+  # GET /articles/1.json
+  def show
+    #increase views(bad implementation as single user can give multiple views)
```

Comment:
Interesting - how you would implement it in a "good way"?

Reply:
I could implement it in such a way that one user can give only one view.

Reply:
Yes but how? Will you add more tables, columns?

Reply:
I can create a new model called `views` (which can also store whether a user liked or disliked an article).
```ruby
create_table :views do |t|
  t.integer :user_id
  t.integer :article_id
  t.boolean :liked, default: false
  t.boolean :disliked, default: false
end
```
Can  you please tell me how association can be implemented here.

Reply:
Alright, so there are two main associations here - An article has many views (different users that have viewed this article) and a user has many articles that they have seen. The most appropriate association is `has_many through`: https://guides.rubyonrails.org/association_basics.html#the-has-many-through-association

Other than that, to avoid counting same views again - you need to ensure the tuple `(user_id, article_id)` is always unique.

Reply:
Thank you for the help